### ISSUE TITLE
Custom middleware

### ISSUE THREAD

**Issue by tdegeus** [sciunto-org/python-bibtexparser#440]
**Is your feature request related to a problem? Please describe.**
It is not completely clear how I can add custom middleware operations.

**Describe the solution you'd like**
This is the simplest API that I would expect:
```python
bibtex_str = """
@ARTICLE{Cesar2013,
  author = {Jean César and T. W. J. de Geus},
  title = {An amazing title},
  year = {2013},
  pages = {12--23},
  volume = {12},
  journal = {Nice Journal}
}
"""

import bibtexparser
import bibtexparser.middlewares as Blockmiddelware

abbreviations = {
    "Nice Journal": "NJ",
    "Another Nice Journal": "ANJ"
}

class JournalAbbreviate(Blockmiddelware):
    def transform_entry(self, entry, *args, **kwargs):
        if entry["journal"] in abbreviations:
            entry["journal"] = abbreviations[entry["journal"]]
        return entry

library = bibtexparser.parse_string(bibtex_str, append_middleware=[JournalAbbreviate()])
print(bibtexparser.write_string(library))
```
but it throws
```
Traceback (most recent call last):
  File "mytest.py", line 20, in <module>
    class JournalAbbreviate(Blockmiddelware):
TypeError: module() takes at most 2 arguments (3 given)
```

**Use cases**
-

**Describe alternatives you've considered**
-

**Remaining Questions (Optional)**
Please tick all that apply:

- [x] I would be willing to contribute a PR to fix this issue.
- [x] This issue is a blocker, I'd be grateful for an early fix.

**Comment by MiWeiss** [comment]
Thanks a lot for the report, that one confused me at first :-)

Turned out there was a simple typo in the docs. You should extend `bibtexparser.middlewares.BlockMiddleware` (note the capital `M`). 

Then, the IDE or shown errors should provide straightforward instructions on what else there is to fix. More docs would certainly help but are not a priority at the moment.

See #444 for the fix, which also checks in your example as a sanity check.

**Comment by tdegeus** [comment]
Thanks. I have a problem though:
```python
import bibtexparser
from bibtexparser.middlewares.middleware import BlockMiddleware

bibtex_str = """
@ARTICLE{Cesar2013,
author = {Jean César and T. W. J. de Geus},
title = {An amazing title},
year = {2013},
pages = {12--23},
volume = {12},
journal = {Nice Journal}
}
"""

abbreviations = {"Nice Journal": "NJ", "Another Nice Journal": "ANJ"}

class JournalAbbreviate(BlockMiddleware):
    def transform_entry(self, entry, *args, **kwargs):
        if entry["journal"] in abbreviations:
            entry["journal"] = abbreviations[entry["journal"]]
        return entry

library = bibtexparser.parse_string(bibtex_str, append_middleware=[JournalAbbreviate(True)])
assert library.entries[0]["journal"] == "NJ"
```
raises
```
raceback (most recent call last):
  File "/Users/tdegeus/data/prog/src/GooseBib/t.py", line 23, in <module>
    library = bibtexparser.parse_string(bibtex_str, append_middleware=[JournalAbbreviate(True)])
                                                                       ^^^^^^^^^^^^^^^^^^^^^^^
TypeError: Can't instantiate abstract class JournalAbbreviate without an implementation for abstract method 'metadata_key'
```
or is this fixed since `2.0.0b5`

**Comment by tdegeus** [comment]
Neve mind, fixed in `2.0.0b6`
